File: Solutions/C#/0050_Pow(x,n)/Pow.cs
using System;

public class Solution
{
    public double MyPow(double x, int n)
    {
        if (x == 1 || n == 0)
        {
            return 1;
        }



        if (n < 0)
        {
            long p = (long)n;
            return 1 / FindPow(x, -p);
        }

        return FindPow(x, n);
    }


    double FindPow(double x, long n)
    {
        if (n == 1)
            return x;

        if (n % 2 == 1)
        {
            return x * FindPow(x * x, (n - 1) / 2);
        }

        return FindPow(x * x, n / 2);
    }
}

File: Solutions/C++/Math/Pow.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;


class Solution {
public:
    // 2 ^ 10 = 4 ^ 5 = 16 ^ 2 * 4 = 256 ^ 1 * 4 = 256 * 4 = 1024
    double myPow(double x, int n) {
        if(n == 0)
            return 1;

        bool isInverse = (n < 0);
        double remainder = 1;
        if(n == INT_MIN)
        {
            remainder *= x;
            n += 1;
        }

        n = abs(n);

        while(n != 1) {
            if(n % 2 == 1) {
                remainder *= x;
                n--;
            }

            n /= 2;
            x *= x;
        }

        x *= remainder;
        if(isInverse)
            x = 1 / x;

        return x;
    }
};